src/components/data-exploration/heatmap/HeatmapPlot.jsx

```diff
@@ -106,29 +139,21 @@ const HeatmapPlot = (props) => {
     setIsHeatmapGenesLoading(false);
   }, [selectedGenes, loadingGenes, markerGenesLoading]);
 
-  useEffect(() => {
-    if (cellSetsLoading || cellSetsHierarchy.length === 0) {
-      return;
-    }
-
-    const legends = legendIsVisible ? spec.legends : [];
-    setVegaSpec({ ...spec, legends });
-  }, [legendIsVisible]);
-
   useEffect(() => {
     if (!selectedGenes?.length > 0
       || cellSetsHierarchy.length === 0
-      || _.isEqual(currentHeatmapSettings, heatmapSettings)
+      || _.isEqual(currentHeatmapSettingsRef, heatmapSettings)
     ) {
       return;
     }
 
-    currentHeatmapSettings.current = heatmapSettings;
+    currentHeatmapSettingsRef.current = heatmapSettings;
 
     const data = populateHeatmapData(
-      cellSets, heatmapSettings, expressionData, selectedGenes, true,
+      cellSets, heatmapSettings, expressionData, selectedGenes, true, true,
     );
-    setVegaDataWithDebounce(data);
+
+    setHeatmapDataWithDebounce(data);
```

Comment:
somewhere here you have a useEffect that gets triggered if either of the following changes:

`  }, [
    selectedGenes,
    heatmapSettings,
    maxCells,
    markerGenesLoading,
    cellSetsLoading,
    cellSetsHidden,
  ]);`

why only depending on `markerGenesLoading`, but not `loadingGenes`?

Reply:
I'll try removing `markerGenesLoading`, because this shoudl change when selected genes data changes (so on `selectedGenes` changing)